app/test/testMetadata.js

```diff
@@ -0,0 +1,161 @@
+  title: 'An Example Page',
+  url: 'http://www.example.com/good'
+};
+
+const goodExampleUrl = 'http://www.example.com/good';
+const badExampleUrl = 'http://www.example.com/bad';
+
+
+describe('Metadata API Tests', function() {
+  before(function() {
+    fetchMock.mock(
+      goodExampleUrl,
+      `
+        <html>
+          <head>
+            <title>${exampleMetadata.title}</title>
+            <meta name="description" content="${exampleMetadata.description}" />
+            <meta name="thumbnail" content="${exampleMetadata.images[0].url}" />
+            <link rel="icon" href="http://www.example.com/rich-icon.png" />
+          </head>
+          <body></body>
+        </html>
+      `
+    );
+    fetchMock.mock(badExampleUrl, 500);
+  });
+
+  after(function() {
+    fetchMock.restore();
+  });
+
+  it('should raise 400 if missing JSON body', (done) => {
+    chai.request(app)
+      .post('/')
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if missing urls parameter', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if urls is empty', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({urls: []}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should raise 400 if content header is not set', (done) => {
+    chai.request(app)
+      .post('/')
+      .send(JSON.stringify({urls: [goodExampleUrl]}))
+      .end((err, res) => {
+        res.should.have.status(400);
+        done();
+      });
+  });
+
+  it('should return 200 with metadata for a url', (done) => {
+    chai.request(app)
+      .post('/')
+      .set('content-type', 'application/json')
+      .send(JSON.stringify({urls: [goodExampleUrl]}))
+      .end((err, res) => {
+
+        res.should.have.status(200);
+
+        const expectedResponse = {
+          error: '',
+          urls: {}
+        };
+        expectedResponse.urls[goodExampleUrl] = exampleMetadata;
+
+        res.body.should.deep.equal(expectedResponse);
+
+        done();
+      });
+  });
+});
+
+describe('Metadata API Tests', function() {
+  before(function() {
+    fetchMock.mock(
+      goodExampleUrl,
+      `
+        <html>
+          <head>
+            <title>${exampleMetadata.title}</title>
+            <meta name="description" content="${exampleMetadata.description}" />
+            <meta name="thumbnail" content="${exampleMetadata.images[0].url}" />
+          </head>
+          <body></body>
+        </html>
+      `
+    );
+  });
+
+  after(function() {
+    fetchMock.restore();
+  });
+
+  it('should fallback to favicon.ico if no rich icon found', (done) => {
+    const expectedMetadata = JSON.parse(JSON.stringify(exampleMetadata));
```

Comment:
Yeah that only does a shallow copy, apparently there isn't a built in way to do a deep copy in node.js.